msal/region.py

```diff
@@ -0,0 +1,33 @@
+import os
+import json
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def _detect_region(http_client=None):
+    region = _detect_region_of_azure_function()  # It is cheap, so we do it always
+    if http_client and not region:
+        return _detect_region_of_azure_vm(http_client)  # It could hang for minutes
+    return region
+
+
+def _detect_region_of_azure_function():
+    return os.environ.get("REGION_NAME")
+
+
+def _detect_region_of_azure_vm(http_client):
```

Comment:
I think this is a little more complicated ... see https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/1770654e7a2985dc06849366e5ae8e229b22faa7/src/client/Microsoft.Identity.Client/Instance/Region/RegionManager.cs#L192

As far as I understood, the algorithm is: 

- try with IDMS api version `2020-06-01`
- if you get a BadRequest (400) error then make a request to `http://169.254.169.254/metadata/instance/compute/location` which will still return a 400 error but the response will have the API version https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/1770654e7a2985dc06849366e5ae8e229b22faa7/src/client/Microsoft.Identity.Client/Instance/Region/LocalImdsErrorResponse.cs#L21
- Use this new api version parameter and retry the call

@neha-bhargava - could you update the API design document with this algorithm please, since you understand it better.

Reply:
The api version used here is recent I believe which is fine. But these api versions keep on updating and after sometime the version we mentioned here might no longer be available. So to handle that we check for the BadRequest error and make a request without the api version param which returns the latest api version which we can use to make the api call.

Reply:
Updated the api doc: https://identitydivision.visualstudio.com/DefaultCollection/DevEx/_git/AuthLibrariesApiReview/pullrequest/3291

Reply:
The approach we selected here was to reduce the possibility of failing when they discontinue the version we have. Hardcoding it to make less network calls and because it is true that this would happen far less and far long. Is there any other approach that you would like to suggest for future to avoid failures when the version we mention is no longer supported?

Reply:
Hardcoding an existing known-to-work version is not even about making less network calls. It is because it is the (only?) right way to do versioning among loosely-coupling web services.

And, nope, there will be no programmatical way to avoid failures like this. There has to be out-of-band (OOB) information to trigger such change. Those OOB info would ideally be conveyed as an ahead notice from IMDS team to us, or - if I may be honest here - via an IcM.